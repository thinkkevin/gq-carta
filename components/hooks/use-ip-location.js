import useSWR from "swr";

export function useIPLocation() {
  const { data, error } = useSWR("check", ip_fetcher);
  return {
    error,
    ...data,
  };
}

const IP_API_URL = "http://api.ipstack.com/check";

async function ip_fetcher() {
  const res = await fetch(
    IP_API_URL + "?access_key=" + process.env.IPSTACK_KEY
  );
  const json = await res.json();
  return {
    city: json.city,
    state: json.region_code,
    latitude: json.latitude,
    longitude: json.longitude,
    country: json.country_name,
    display: `${json.city}, ${json.region_code}`,
  };
}
