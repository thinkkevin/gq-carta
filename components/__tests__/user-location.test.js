import CurrentLocation, {
  UserLocationContext,
  IPLocation,
  UserLocation,
} from "../user-location";

import * as useHooks from "../hooks/use-ip-location";

import { shallow } from "enzyme";
describe("Test user-location components", () => {
  it("verify UserLocation Component", () => {
    let comp = (
      <UserLocation>
        <UserLocationContext.Consumer>
          {(location) => {
            return (
              <div>
                {location.display}({location.latitude}, {location.longitude})
              </div>
            );
          }}
        </UserLocationContext.Consumer>
      </UserLocation>
    );
    let html = shallow(comp).html();
    expect(html).toContain("Current Location"); // use "Current Location" as hard-coded text when using UserLocation
    expect(html).toMatchSnapshot();
  });

  it("verify IPLocation using result from useIPLocation", () => {
    let mock = jest.spyOn(useHooks, "useIPLocation");
    mock.mockImplementation(() => {
      return {
        latitude: 37.55649948120117,
        longitude: -121.97046661376953,
        display: "Fremont, CA",
        city: "Fremont",
        state: "CA",
        country: "United States",
        error: null,
      };
    });
    let comp = (
      <IPLocation>
        <UserLocationContext.Consumer>
          {(location) => {
            return (
              <div>
                {location.display}({location.latitude}, {location.longitude})
              </div>
            );
          }}
        </UserLocationContext.Consumer>
      </IPLocation>
    );

    let html = shallow(comp).html();
    expect(html).toContain("Fremont, CA"); // use "Current Location" as hard-coded text when using UserLocation
    expect(html).toMatchSnapshot();
    mock.mockRestore();
  });

  it("verify CurrentLocation Component will use UserLocation first", () => {
    let comp = (
      <CurrentLocation>
        <UserLocationContext.Consumer>
          {(location) => {
            return (
              <div>
                {location.display}({location.latitude}, {location.longitude})
              </div>
            );
          }}
        </UserLocationContext.Consumer>
      </CurrentLocation>
    );
    let html = shallow(comp).html();
    expect(html).toContain("Current Location"); // use "Current Location" as hard-coded text when using UserLocation
    expect(html).toMatchSnapshot();
  });
});
