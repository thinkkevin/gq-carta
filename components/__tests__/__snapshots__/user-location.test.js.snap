// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test user-location components verify CurrentLocation Component will use UserLocation first 1`] = `"<div>Current Location(37.5751118299, -122.0408938)</div>"`;

exports[`Test user-location components verify IPLocation using result from useIPLocation 1`] = `"<div>Fremont, CA(37.55649948120117, -121.97046661376953)</div>"`;

exports[`Test user-location components verify UserLocation Component 1`] = `"<div>Current Location(37.5751118299, -122.0408938)</div>"`;
